### PR TITLE
blobstore: wire SSE-KMS into Alibaba (OSS) async directory upload

### PR DESCRIPTION
**Pull request by kchoy-sfdc**
## Summary

Wires SSE-KMS into the Alibaba (OSS) async directory-upload path, bringing it to parity with single-object OSS upload (which already supports KMS) and with the AWS/GCP directory-upload KMS support added in #598.

## Changes

- `AliAsyncBlobStore.doUploadDirectory` now propagates `kmsKeyId` and `useKmsManagedKey` from the `DirectoryUploadRequest` onto each per-file `UploadRequest`. `AliTransformer.toPutObjectRequest` already maps these to OSS SSE-KMS (explicit `kmsKeyId` takes precedence; otherwise `useKmsManagedKey` selects a managed CMK), so no transformer change is needed.
- Sync `AliBlobStore` directory upload remains `UnsupportedOperationException` (unchanged; Ali directory upload is async-only).
- Unit tests (`AliAsyncBlobStoreTest`): capture each per-file `PutObjectRequest` and assert SSE-KMS propagation for the explicit-key, managed-key, and no-KMS cases. The explicit-key test uses a file nested under a subdirectory and asserts the per-file object keys (`prefix/file1.txt`, `prefix/subdir/file2.txt`) to pin key/prefix handling alongside KMS propagation.

## Testing

- `mvn test -pl blob/blob-ali` — all unit tests pass.
- Verified end-to-end against live cn-shanghai OSS: directory upload with an explicit KMS key, a managed key, and no key produced the expected `x-oss-server-side-encryption` headers on every uploaded object, including files under subdirectories.

## Cross-cloud

- Depends on the `DirectoryUploadRequest` KMS fields added for AWS/GCP in #598 (merged).
- Ali async directory upload now matches AWS/GCP; sync Ali directory upload remains unsupported (async-only, unchanged).

### CODE REVIEW

File: blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java [hunk 1]
```diff
@@ -1145,6 +1145,114 @@ void testUploadDirectory(@TempDir Path tempDir) throws Exception {
...
+    when(mockAsyncClient.putObjectAsync(
+        any(PutObjectRequest.class), any(OperationOptions.class)))
+        .thenReturn(CompletableFuture.completedFuture(mockResult));
+
+    String kmsKeyId = "acs:kms:cn-shanghai:1099202394511537:key/test-key";
+    DirectoryUploadRequest request = DirectoryUploadRequest.builder()
+        .localSourceDirectory(tempDir.toString())
+        .prefix("kms-prefix")
+        .includeSubFolders(true)
+        .kmsKeyId(kmsKeyId)
```
Comment (LihaoLiuXs) [inline blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java:1168]: Please exercise the both-flags case here by also setting `useKmsManagedKey(true)`. This PR promises that the explicit key wins, but these directory tests only cover mutually exclusive inputs. I changed the new propagation to drop `kmsKeyId` whenever the managed flag is true; all three added tests still passed, so a directory request with both values could silently lose its caller-selected key. Adding the flag here made this test fail with the expected key ID versus `null`. The existing transformer precedence test starts from an `UploadRequest`, so it cannot catch a regression in the directory-request propagation.

Comment (iamabhilaksh) [reply blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java:1168]: +1 — worth locking down the both-flags precedence (explicit `kmsKeyId` + `useKmsManagedKey(true)`) with a test before this merges; the directory propagation path isn't otherwise pinned.

**Review by iamabhilaksh — COMMENTED** [review]
Holding approval on the open inline item — will re-review once it's covered. Rest of the change looks good: clean SSE-KMS parity with the other substrates, and the negative (no-KMS) case is a nice touch.

File: blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java (hunk 1 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java:1168]: Good catch — thanks for the mutation test. You're right that the three cases were all mutually-exclusive, so the both-flags precedence wasn't pinned at the directory-propagation level (and the transformer precedence test can't cover it since it starts from a hand-built `UploadRequest`). Fixed in `6351c0bfb`: added `testUploadDirectoryExplicitKmsKeyWinsOverManagedKey`, which sets **both** `kmsKeyId` and `useKmsManagedKey(true)` and asserts each per-file `PutObjectRequest` carries the explicit key id (explicit wins). I verified it's a real discriminator by locally applying your exact mutant (drop `kmsKeyId` when the managed flag is set) — only this new test went red; the mutually-exclusive cases survived, matching what you saw.

To be clear on current behavior: production code is unaffected — `doUploadDirectory` propagates both fields verbatim and `AliTransformer.toPutObjectRequest` enforces precedence (explicit `kmsKeyId` wins over `useKmsManagedKey`), so the explicit key is never dropped. This is purely closing the test-coverage gap.

Comment (kchoy-sfdc) [reply blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java:1168]: Done — added `testUploadDirectoryExplicitKmsKeyWinsOverManagedKey` in `6351c0bfb`, which pins exactly that: both `kmsKeyId` and `useKmsManagedKey(true)` set, asserting each per-file `PutObjectRequest` keeps the explicit key id. Confirmed it catches the directory-propagation regression by applying the drop-`kmsKeyId`-when-managed mutant locally — only this new test failed. Thanks both for flagging it.

**Review by iamabhilaksh — APPROVED** [review]
LGTM 👍

The follow-up commit closes the gap cleanly — the propagation path was always explicit-key-wins, and the new both-flags test pins that precedence at exactly the layer that was previously unguarded. The KMS test matrix is now complete (key-only / managed-only / both / neither).

File: blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java (hunk 1 above)
Comment (iamabhilaksh) [reply blob/blob-ali/src/test/java/com/salesforce/multicloudj/blob/ali/async/AliAsyncBlobStoreTest.java:1168]: Confirmed — 6351c0bf adds the both-flags test that pins explicit-key-wins, which is what this thread was after. Nice.

**Review by LihaoLiuXs — APPROVED** [review]
